cmd/gwi-api/api/api_test.go

```diff
@@ -0,0 +1,96 @@
+package api_test
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api"
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api/stores/mockstore"
+
+	"github.com/labstack/echo"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"go.uber.org/zap"
+)
+
+// ----------------------------------------------------------------------------
+// NOTE: Testing a single enpoint for time purposes only.
+// ----------------------------------------------------------------------------
+
+func TestAPI_GetTeamByID(t *testing.T) {
+	assert := assert.New(t)
+
+	// Testing endpoints returning *echo.HTTPError is kinda weird.
+	// You cannot check the response object, you have to check the error object being returned.
+	// So in case ExpectedResponseError is set, we do not check ExpectedResponseStatusCode at all.
+	testCases := []struct {
+		RequestedTeamID            string
+		StoredTeam                 *api.Team
+		ExpectedResponseError      error
+		ExpectedResponseStatusCode int
+		ExpectedResponseBody       echo.Map
+	}{
+		// Team found
+		{
+			RequestedTeamID: "gladiators",
+			StoredTeam: &api.Team{
+				ID:   "gladiators",
+				Name: "The Gladiators",
+			},
+			ExpectedResponseStatusCode: http.StatusOK,
+			ExpectedResponseBody: echo.Map{
+				"id":   "gladiators",
+				"name": "The Gladiators",
+			},
+		},
+		// Team not found
+		{
+			RequestedTeamID:       "gladiators",
+			ExpectedResponseError: echo.ErrNotFound,
+		},
+	}
+
+	for _, tc := range testCases {
+		// Mock the store.
+		store := &mockstore.Store{}
```

Comment:
Can you see any pitfalls in writing tests using a store that's potentially different from the store used in production?

Reply:
I can see pitfalls in any tests that are not mimicking production deployment exactly 🙂 On the other hand, these are unit tests. You are expected to mock what the piece of code is calling. That is why I am using an interface here, so that it is easy to mock it in tests. Testing the integration is for integration, black box or end-to-end tests...

Reply:
Having said that, it is used for including migration SQL files in the executable. Since we are using Docker it is not that difficult to just add these SQL files into the image and find them on the disk.

Reply:
I believe the second answer rather belongs to [this thread](https://github.com/tchap/gwi-challenge/pull/1#discussion_r282455450) right? In fact that's what we do in our projects.

Reply:
About those tests, I think a hundred people would have a hundred different opinions about that. Our approach is that we mock away only 3rd party tools during tests. That allows us to write only a single layer of tests covering the outer API (HTTP, pubsub handlers, external APIs, 3rd party integrations, ...) - the contract with the outside world - and that means all tests are performed on real databases, queues and so on. It makes the test setup slightly more complicated (but not that bad really), but in the end the tests are very close to reality. 

For example we don't write tests like [this](https://github.com/tchap/gwi-challenge/pull/1/files#diff-43dccfb1085296715c3b58ac56a76569R36) because it's really easy to make a mistake in SQL which the tests wouldn't catch. Instead we assert that the database returns correct data.

Reply:
Yeah, I was also writing tests like that in the previous work. Just writing unit tests for pure stuff, like encoding something and stuff. I agree that when you have nicely separated black boxes with a clean API, this way of testing can be enough. Rewriting unit and integration tests all the time just because you changed internals and not the API is a major PITA I have to say. Testing interfaces FTW.